src/librustc_mir/borrow_check/mod.rs

```diff
@@ -740,6 +740,30 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                     self.visit_terminator_drop(loc, term, flow_state, &place, field_ty, span);
                 }
             }
+            // Same as above, but for tuples.
+            ty::TyTuple(tys) => {
+                for (index, field) in tys.iter().enumerate() {
+                    let gcx = self.tcx.global_tcx();
+                    let field_ty = gcx.normalize_erasing_regions(self.param_env, *field);
+                    let place = drop_place.clone().field(Field::new(index), field_ty);
+
+                    self.visit_terminator_drop(loc, term, flow_state, &place, field_ty, span);
+                }
+            }
+            // Closures and generators also have disjoint fields, but they are only
+            // directly accessed in the body of the closure/generator.
+            ty::TyClosure(def, substs)
+            | ty::TyGenerator(def, substs, ..)
```

Comment:
Generators are a bit funny -- I'm not sure if `upvar_tys` is correct for that case? 

Maybe @Zoxc can comment how to get the list of fields for a generator type, else I'll dig in later.